src/Adapter/Product/CommandHandler/UpdateProductBasicInformationHandler.php

```diff
@@ -108,6 +118,15 @@ private function fillUpdatableProperties(
             $updatableProperties['name'] = array_keys($localizedNames);
         }
 
+        foreach ($product->link_rewrite as $langId => $linkRewrite) {
```

Comment:
`link_rewrite` is multilang field, so it can be string or array ?

Reply:
it is always array of strings indexed by langId

Reply:
@zuk3975 please see Product `link_rewrite` attr definition.
https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Product.php#L173

Reply:
yep i see and it is wrong. Lang fields always returns array.
I'll fix the typehint to be string[] and hopefully phpstan won't complain

Reply:
okay so I've changed it and we see some errors from other parts of code.
I personally haven't faced a scenario when $product->link_rewrite is a string (so im hoping that the typehints are messed up in legacy), but wdyt?
Is it possible that in some cases is string (like in some shown in phpstan error?)
If its possible then how the hell we suppose to deal with it? (I think in whole product page migration with multilang values we assumed its always array :thinking: )
ping @jolelievre @PululuK

Reply:
Ok so apparently it can be a string (thanks @jolelievre for explanation - `if you specify the $langId parameter in the constructor then name field will be a string picked from the localized arra and matching the specified language`), but in this case (and whole product page migration) we always load full product and expect localized array everywhere.

Reply:
So `link_rewrite` actually behaves the same as any multilang field in any ObjectModel:
- if you provide langId in the constructor only one value from the localized array is kept and its type is string
- if you don't provide any langId then it keeps the localized array value

But in all our handlers we use the same repository which never indicates the `langId` since we are in an edition context we always need to get the full array for all languages, therefore in this context it's an array not a string. Adding additional checks everywhere in all our handlers would be counterproductive and would reduce the code readability for nothing.

I'm surprised you didn't know this @PululuK :trollface:

Reply:
The product here is the legacy ObjectModel, so how do you know Language ID was set during object instantiation?
```php
UpdateProductSeoHandler::fillUpdatableProperties(Product $product, UpdateProductSeoCommand $command): array
```

About :
> But in all our handlers we use the same repository which never indicates the langId since we are in an edition context we always need to get the full array for all languages, therefore in this context it's an array not a string. Adding additional checks everywhere in all our handlers would be counterproductive and would reduce the code readability for nothing.

Well, in this case we have to use a factory, instead of expecting a Product Object model, we should use an entity ID in params. It's much more secure. what will happen if someone forgets and indicates the Id language?

Reply:
its a private method called just after product is retrieved, I don't think we should worry about it

Reply:
@PululuK that's already what we do when we use this `$product = $this->productRepository->getByShopConstraint($command->getProductId(), $command->getShopConstraint());` All the command handlers are based on this repository and this repository NEVER indicates the `languageID` That's how we know we always have a "multi lang" product here

And if someone messes with the repository every behat tests will start failing so I don't see how it could be modified